pczt/src/roles/signer/mod.rs

```diff
@@ -0,0 +1,311 @@
+            .recipient()
+            .zip(spend.value().as_ref())
+            .zip(spend.rseed().as_ref())
+            .map(|((recipient, value), rseed)| {
+                sapling::Note::from_parts(recipient, *value, *rseed)
+            });
+
+        if let Some(note) = note_from_fields {
+            let tx_spend = self
+                .tx_data
+                .sapling_bundle()
+                .expect("index checked above")
+                .shielded_spends()
+                .get(index)
+                .expect("index checked above");
+
+            let proof_generation_key = spend
+                .proof_generation_key()
+                .as_ref()
+                .ok_or(Error::MissingProofGenerationKey)?;
+
+            let nk = proof_generation_key.to_viewing_key().nk;
+
+            let merkle_path = spend.witness().as_ref().ok_or(Error::MissingWitness)?;
+
+            if &note.nf(&nk, merkle_path.position().into()) != tx_spend.nullifier() {
+                return Err(Error::InvalidNullifier);
+            }
+        }
+
+        spend
+            .sign(self.shielded_sighash, ask, OsRng)
+            .map_err(Error::SaplingSign)?;
+
+        Ok(())
+    }
+
+    /// Signs the Orchard spend at the given index with the given spend authorizing key.
+    ///
+    /// It is the caller's responsibility to perform any semantic validity checks on the
+    /// PCZT (for example, comfirming that the change amounts are correct) before calling
+    /// this method.
+    pub fn sign_orchard(
+        &mut self,
+        index: usize,
+        ask: &orchard::keys::SpendAuthorizingKey,
+    ) -> Result<(), Error> {
+        let action = self
+            .orchard
+            .actions_mut()
+            .get_mut(index)
+            .ok_or(Error::InvalidIndex)?;
+
+        // Check consistency of the input being signed.
+        let note_from_fields = action
+            .spend()
+            .recipient()
+            .zip(action.spend().value().as_ref())
+            .zip(action.spend().rho().as_ref())
+            .zip(action.spend().rseed().as_ref())
+            .map(|(((recipient, value), rho), rseed)| {
+                orchard::Note::from_parts(recipient, *value, *rho, *rseed)
+                    .into_option()
+                    .ok_or(Error::InvalidNote)
+            })
+            .transpose()?;
+
+        if let Some(note) = note_from_fields {
+            let tx_action = self
+                .tx_data
+                .orchard_bundle()
+                .expect("index checked above")
+                .actions()
+                .get(index)
+                .expect("index checked above");
+
+            let fvk = action
+                .spend()
+                .fvk()
+                .as_ref()
+                .ok_or(Error::MissingFullViewingKey)?;
+
+            if &note.nullifier(fvk) != tx_action.nullifier() {
+                return Err(Error::InvalidNullifier);
+            }
+        }
+
```

Comment:
```suggestion

        // The sighash for an Orchard spend depends on all effecting data.
        self.tx_modifiable &= !(FLAG_INPUTS_MODIFIABLE | FLAG_OUTPUTS_MODIFIABLE);

```
(If `action.sign` is guaranteed to be non-mutating when it fails, this could be after the call.)